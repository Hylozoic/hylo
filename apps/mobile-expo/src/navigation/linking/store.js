import { create } from 'zustand'

const useLinkingStore = create((set) => ({
  initialURL: null,
  setInitialURL: (url) => set({ initialURL: url }),
  returnToOnAuthPath: null,
  setReturnToOnAuthPath: (path) => set({ returnToOnAuthPath: path })
}))

export default useLinkingStore