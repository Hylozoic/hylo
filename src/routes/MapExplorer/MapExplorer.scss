.container {
  position: relative;
  height: calc(100vh - 56px);
  overflow: hidden;

  &.noUser {
    height: 100%;
    width: 100%;
  }
}

.expanded {
  margin: 0 $space-2x $space-6x;
}

.toggleDrawerButton {
  position: fixed;
  right: 20px;
  top: 75px;
  border-radius: 20px;
  width: 40px;
  height: 40px;
  background: white;
  transition: all .25s ease;
  box-shadow: 0px 4px 15px rgba(0, 0, 0, 0.15);
  border: none;
  text-align: center;
  display: block;

  &.drawerOpen {
    right: calc(40% + 20px);

    .closeDrawer {
      display: block;
      position: relative;
      top: 2px;
      width: 23px;
      height: 23px;
      position: absolute;
      top: 50%;
      left: 50%;
      transform: translate(-50%, -50%);
    }

    .openDrawer {
      display: none;
    }
  }

  &:hover {
    cursor: pointer;
  }

  .openDrawer {
    font-size: 16px;
    position: relative;
    top: 2px;
    display: block;
    width: 20px;
    height: 16px;
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
  }

  .closeDrawer {
    display: none;
  }
}

.noUser .toggleDrawerButton {
  top: 20px;
}

@media screen and (max-width: 1115px) {
  .toggleDrawerButton.drawerOpen {
    right: 470px;
  }
}

.toggleDrawerButton:focus {
  outline: 0;
}

.icon {
  vertical-align: middle;
  text-align: center;
}

.postTip {
  position: absolute;
  z-index: 1;
  pointerEvents: none;
  border-radius: 4px;
  background-color: white;
  padding: 0 15px;
  color: rgba(87, 102, 123, 1.000);
  box-shadow: 0px 4px 15px rgba(0, 0, 0, 0.15);
  line-height: 28px;

  p {
    margin: 0;
    padding: 0;
  }
}

.postTip.discussion {
  color: $color-discussions;
}

.postTip.event {
  color: $color-events;
}

.postTip.offer {
  color: $color-offers;
}

.postTip.resource {
  color: $color-resources;
}

.postTip.project {
  color: $color-projects;
}

.postTip.cluster {
  color: $color-black;
}

button.toggleFeatureFiltersButton {
  position: absolute;
  bottom: 10px;
  left: 20px;
  color: rgba(125, 140, 157, 1.000);
  background-color: rgba(255,255,255,1);
  border: none;
  outline: none !important;
  border-radius: 5px;
  height: 30px;
  text-align: center;
  box-shadow: 0px 4px 15px rgba(0, 0, 0, 0.15);
  padding: 0 10px;

  &.featureFiltersOpen {
    color: rgba(255,255,255,1);
    background-color: $color-caribbean-green;
  }
}

.searchAutocomplete {
  position: absolute;
  top: 25px;
  left: 20px;
  color: rgba(125, 140, 157, 1.000);
  background-color: rgba(255, 255, 255, 1);
  border: none;
  outline: none !important;
  border-radius: 5px;
  height: 30px;
  text-align: center;
  box-shadow: 0px 4px 15px rgba(0, 0, 0, 0.15);
  padding: 0 10px;
  padding-top: 4px;
}

.pointer {
  width: 10px;
  height: 10px;
  transform: rotate(45deg);
  position: absolute;
  bottom: -5px;
  left: 55px;
  background-color: white;
}

.featureTypeFilters {
  position: absolute;
  bottom: 50px;
  left: 20px;
  background: white;
  display: none;
  padding-bottom: 10px;
  box-shadow: 0px 4px 15px rgba(0, 0, 0, 0.15);

  &.featureFiltersOpen {
    display: block;
  }

  h3 {
    font-size: 14px;
    margin-bottom: 15px;
    padding: 10px;
    border-bottom: 1px solid rgba(224, 229, 233, 1.000);
  }
}

.featureTypeSwitch {
  margin: 10px;
  > span {
    margin-left: 10px;
  }
}


@media screen and (max-width: 600px) {
  .toggleDrawerButton.drawerOpen {
    left: 10px;
    right: auto;
  }

  .featureTypeFilters {
    bottom: 106px;
  }

  button.toggleFeatureFiltersButton {
    bottom: 66px;
  }
}

@media screen and (max-width: 375px) {
  .toggleDrawerButton.drawerOpen {
    z-index: 100;
  }
}