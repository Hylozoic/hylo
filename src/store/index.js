import { createStore } from 'redux'
import createMiddleware from './middleware'
import { createBrowserHistory } from 'history'
import createRootReducer, { createCombinedReducers } from './reducers'

// `window` check here is for SSR case (see `appMiddleware`)
// where a static memory history is passed
export const history = typeof window !== 'undefined' && createBrowserHistory()

export function getEmptyState (providedHistory = history) {
  const combinedReducers = createCombinedReducers(providedHistory)

  return combinedReducers({}, { type: '' })
}

export default function configureStore (providedHistory = history, req) {
  return createStore(
    createRootReducer(providedHistory),
    getEmptyState(providedHistory),
    createMiddleware(providedHistory, req)
  )
}
