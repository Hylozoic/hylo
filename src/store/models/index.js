import { ORM } from 'redux-orm'
import './Model.extension'
import Activity from './Activity'
import Attachment from './Attachment'
import Comment from './Comment'
import EventInvitation from './EventInvitation'
import Group, { GroupConnection, GroupModerator, GroupJoinQuestion } from './Group'
import GroupTopic from './GroupTopic'
import Invitation from './Invitation'
import JoinRequest, { JoinRequestQuestionAnswer, Question } from './JoinRequest'
import LinkPreview from './LinkPreview'
import Location from './Location'
import Me, { MySkillsToLearn } from './Me'
import Membership from './Membership'
import Message from './Message'
import MessageThread from './MessageThread'
import Notification from './Notification'
import Person, { PersonSkillsToLearn } from './Person'
import PersonConnection from './PersonConnection'
import Post, { PostFollower, PostCommenter, ProjectMember } from './Post'
import PostMembership from './PostMembership'
import SearchResult from './SearchResult'
import Skill from './Skill'
import Topic from './Topic'
import Vote from './Vote'

export const orm = new ORM({ stateSelector: state => state.orm })

orm.register(
  Activity,
  Attachment,
  Comment,
  EventInvitation,
  Group,
  GroupConnection,
  GroupModerator,
  GroupJoinQuestion,
  GroupTopic,
  Invitation,
  LinkPreview,
  Location,
  JoinRequest,
  JoinRequestQuestionAnswer,
  Me,
  Membership,
  Message,
  MessageThread,
  MySkillsToLearn,
  Notification,
  Person,
  PersonConnection,
  PersonSkillsToLearn,
  Post,
  PostCommenter,
  PostFollower,
  PostMembership,
  ProjectMember,
  Question,
  SearchResult,
  Skill,
  Topic,
  Vote
)

export default orm
