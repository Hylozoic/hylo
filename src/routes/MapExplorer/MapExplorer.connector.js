import { push } from 'connected-react-router'
import { connect } from 'react-redux'
import { get, pick } from 'lodash/fp'
import { FETCH_POSTS_MAP } from 'store/constants'
import getRouteParam from 'store/selectors/getRouteParam'
import {
  fetchSavedSearches, deleteSearch
} from '../UserSettings/UserSettings.store'
import getQuerystringParam from 'store/selectors/getQuerystringParam'
import presentPost from 'store/presenters/presentPost'
import getCommunityForCurrentRoute from 'store/selectors/getCommunityForCurrentRoute'
import getMe from 'store/selectors/getMe'
import { addQuerystringToPath, baseUrl, personUrl, postUrl, communityMapDetailUrl } from 'util/navigation'

import {
  fetchMembers,
  fetchPosts,
  fetchPublicCommunities,
  getPublicCommunities,
  getSortedFilteredMembers,
  getSortedFilteredPosts,
  getCurrentTopics,
  storeFetchParams,
  storeClientFilterParams
} from './MapExplorer.store.js'

export function presentMember (person, communityId) {
  return {
    ...pick([ 'id', 'name', 'avatarUrl', 'locationObject', 'tagline', 'skills' ], person.ref),
    type: 'member',
    skills: person.skills.toModelArray(),
    community: person.memberships.first()
      ? person.memberships.first().community.name : null
  }
}

export function mapStateToProps (state, props) {
  const community = getCommunityForCurrentRoute(state, props)
  const communityId = community && community.id
  const routeParams = get('match.params', props)
  const slug = getRouteParam('slug', state, props)
  const context = getRouteParam('context', state, props)
  const networkSlug = getRouteParam('networkSlug', state, props)
  const networkSlugs = getQuerystringParam('network', state, props)
  const hideDrawer = getQuerystringParam('hideDrawer', state, props) === 'true'

  var subject
  if (context === 'public') {
    subject = 'public'
  } else if (slug) {
    subject = 'community'
  } else if (networkSlug) {
    subject = 'network'
  } else {
    subject = 'all'
  }

  const fetchParams = {
    subject,
    slug,
    networkSlug,
    networkSlugs, // networkSlug used by network posts query, networkSlugs uses by public posts query
    boundingBox: state.MapExplorer.fetchParams ? state.MapExplorer.fetchParams.boundingBox : null,
    isPublic: subject === 'public' // only needed to track which query results to pull for the map
  }

  // TODO: maybe filtering should happen on the presentedPosts? since we do some of that presentation in the filtering code, like calling topics.toModelArray in the filters for every post each time
  const members = getSortedFilteredMembers(state, fetchParams).map(m => presentMember(m, communityId))
  const posts = getSortedFilteredPosts(state, fetchParams).map(p => presentPost(p, communityId))
  const topics = getCurrentTopics(state, fetchParams)
  const publicCommunities = getPublicCommunities(state, fetchParams)

  const me = getMe(state)
  const centerLocation = community && community.locationObject ? community.locationObject.center : me && me.locationObject ? me.locationObject.center : null

  return {
    centerLocation: centerLocation || { lat: 35.442845, lng: 7.916598 },
    currentUser: me,
    fetchParams,
    filters: state.MapExplorer.clientFilterParams,
    members,
    pending: state.pending[FETCH_POSTS_MAP],
    posts,
    publicCommunities,
    routeParams,
    hideDrawer,
    searches: state.SavedSearches.searches,
    topics,
    zoom: centerLocation ? 10 : 0
  }
}

export function mapDispatchToProps (dispatch, props) {
  const routeParams = get('match.params', props)
  const querystringParams = getQuerystringParam(['hideDrawer', 't', 'network'], null, props)
  return {
    fetchMembers: (params) => () => dispatch(fetchMembers({ ...params })),
    fetchPosts: (params) => () => dispatch(fetchPosts({ ...params })),
    fetchPublicCommunities: (params) => () => dispatch(fetchPublicCommunities({ ...params })),
    fetchSavedSearches: (userId) => () => dispatch(fetchSavedSearches(userId)),
    deleteSearch: (searchId) => dispatch(deleteSearch(searchId)),
    showDetails: (postId) => dispatch(push(postUrl(postId, { ...routeParams, view: 'map' }, querystringParams))),
    showCommunityDetails: (communityId) => dispatch(push(communityMapDetailUrl(communityId, { ...routeParams, view: 'map' }, querystringParams))),
    gotoMember: (memberId) => dispatch(push(personUrl(memberId, routeParams.slug, routeParams.networkSlug))),
    toggleDrawer: (hidden) => dispatch(push(addQuerystringToPath(baseUrl({ ...routeParams, view: 'map' }), { ...querystringParams, hideDrawer: hidden }))),
    storeFetchParams: param => opts => dispatch(storeFetchParams({ ...param, ...opts })),
    storeClientFilterParams: params => dispatch(storeClientFilterParams(params))
  }
}

export function mergeProps (stateProps, dispatchProps, ownProps) {
  const { fetchParams, currentUser } = stateProps
  const { fetchMembers, fetchPosts, fetchPublicCommunities, storeFetchParams, fetchSavedSearches } = dispatchProps

  return {
    ...ownProps,
    ...stateProps,
    ...dispatchProps,
    fetchMembers: fetchMembers(fetchParams),
    fetchPublicCommunities: fetchPublicCommunities(fetchParams),
    fetchPosts: fetchPosts(fetchParams),
    fetchSavedSearches: fetchSavedSearches(currentUser.id),
    storeFetchParams: storeFetchParams(fetchParams)
  }
}

export default connect(mapStateToProps, mapDispatchToProps, mergeProps)
