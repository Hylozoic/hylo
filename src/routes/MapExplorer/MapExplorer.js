import bbox from '@turf/bbox'
import bboxPolygon from '@turf/bbox-polygon'
import booleanWithin from '@turf/boolean-within'
import center from '@turf/center'
import combine from '@turf/combine'
import { featureCollection, point } from '@turf/helpers'
import React from 'react'
import { FlyToInterpolator } from 'react-map-gl'
import { debounce, get, groupBy, isEqual } from 'lodash'
import cx from 'classnames'
import Icon from 'components/Icon'
import Loading from 'components/Loading'
import { FEATURE_TYPES } from './MapExplorer.store'
import Map from 'components/Map/Map'
import MapDrawer from './MapDrawer'
import SavedSearches from './SavedSearches'
import { createIconLayerFromPostsAndMembers } from 'components/Map/layers/clusterLayer'
import { createIconLayerFromCommunities } from 'components/Map/layers/iconLayer'
import { heart } from 'util/assets'
import SwitchStyled from 'components/SwitchStyled'
import styles from './MapExplorer.scss'
import LocationInput from 'components/LocationInput'
import { locationObjectToViewport } from 'util/geo'
import { isPublicPath, isAllCommunitiesPath, isNetworkPath, isCommunityPath, getCommunitySlugInPath as getCommunitySlug, getNetworkSlugInPath as getNetworkSlug } from 'util/navigation'

export default class MapExplorer extends React.Component {
  static defaultProps = {
    centerLocation: { lat: 35.442845, lng: 7.916598 },
    filters: {},
    members: [],
    posts: [],
    publicCommunities: [],
    routeParams: {},
    hideDrawer: false,
    topics: [],
    zoom: 0
  }

  constructor (props) {
    super(props)
    this.state = {
      clusterLayer: null,
      communityIconLayer: null,
      currentBoundingBox: null,
      features: [],
      hideDrawer: props.hideDrawer,
      hoveredObject: null,
      pointerX: 0,
      pointerY: 0,
      selectedObject: null,
      showFeatureFilters: false,
      totalLoadedBoundingBox: null,
      viewport: {
        latitude: parseFloat(props.centerLocation.lat),
        longitude: parseFloat(props.centerLocation.lng),
        zoom: props.zoom,
        bearing: 0,
        pitch: 0
      }
    }
  }

  componentDidMount () {
    this.refs = {}
    Object.keys(FEATURE_TYPES).forEach(featureType => {
      this.refs[featureType] = React.createRef()
    })

    this.props.fetchSavedSearches();
  }

  componentDidUpdate (prevProps) {
    if (!prevProps) return

    const {
      fetchMembers,
      fetchPosts,
      fetchParams,
      fetchPublicCommunities,
      members,
      posts,
      publicCommunities
    } = this.props

    if (!isEqual(prevProps.fetchParams, fetchParams)) {
      fetchMembers()
      fetchPosts()
      fetchPublicCommunities()
    }

    if (this.state.currentBoundingBox &&
        (!isEqual(prevProps.fetchParams, fetchParams) ||
         !isEqual(prevProps.posts, posts) ||
         !isEqual(prevProps.members, members) ||
         !isEqual(prevProps.publicCommunities, publicCommunities))) {
      this.setState(this.updatedMapFeatures(this.state.currentBoundingBox))
    }
  }

  updatedMapFeatures (boundingBox) {
    const bbox = bboxPolygon(boundingBox)
    const viewMembers = this.props.members.filter(member => {
      const locationObject = member.locationObject
      if (locationObject) {
        const centerPoint = point([locationObject.center.lng, locationObject.center.lat])
        return booleanWithin(centerPoint, bbox)
      }
      return false
    })
    const viewPosts = this.props.posts.filter(post => {
      const locationObject = post.locationObject
      if (locationObject) {
        const centerPoint = point([locationObject.center.lng, locationObject.center.lat])
        return booleanWithin(centerPoint, bbox)
      }
      return false
    })
    const viewCommunities = this.props.publicCommunities.filter(community => {
      const locationObject = community.locationObject
      if (locationObject) {
        const centerPoint = point([locationObject.center.lng, locationObject.center.lat])
        return booleanWithin(centerPoint, bbox)
      }
      return false
    })

    // TODO: update the existing layers instead of creating a new ones?
    return {
      clusterLayer: createIconLayerFromPostsAndMembers({
        members: viewMembers,
        posts: viewPosts,
        onHover: this.onMapHover,
        onClick: this.onMapClick,
        boundingBox: this.state.currentBoundingBox
      }),
      communityIconLayer: createIconLayerFromCommunities({
        communities: viewCommunities,
        onHover: this.onMapHover,
        onClick: this.onMapClick,
        boundingBox: this.state.currentBoundingBox
      }),
      currentBoundingBox: boundingBox,
      features: viewPosts.concat(viewMembers)
    }
  }

  handleLocationInputSelection = (value) => {
    if (value.mapboxId) {
      this.setState({ viewport: locationObjectToViewport(this.state.viewport, value) })
    }
  }

  mapViewPortUpdate = (update) => {
    this.setState({ viewport: update })
  }

  afterViewportUpdate = (update, mapRef) => {
    if (mapRef) {
      let bounds = mapRef.getBounds()
      bounds = [bounds._sw.lng, bounds._sw.lat, bounds._ne.lng, bounds._ne.lat]
      this.updateBoundingBoxQuery(bounds)
    }
  }

  updateBoundingBoxQuery = debounce((newBoundingBox) => {
    let finalBbox
    if (this.state.totalLoadedBoundingBox) {
      const curBbox = bboxPolygon(this.state.totalLoadedBoundingBox)
      const newBbox = bboxPolygon(newBoundingBox)
      const fc = featureCollection([curBbox, newBbox])
      const combined = combine(fc)
      finalBbox = bbox(combined)
    } else {
      finalBbox = newBoundingBox
    }

    // Check if we need to look for more posts and communities
    if (!isEqual(finalBbox, this.state.totalLoadedBoundingBox)) {
      this.props.storeFetchParams({ boundingBox: finalBbox })
    }

    this.setState({
      ...this.updatedMapFeatures(newBoundingBox),
      totalLoadedBoundingBox: finalBbox
    })
  }, 300)

  onMapHover = (info) => this.setState({ hoveredObject: info.objects || info.object, pointerX: info.x, pointerY: info.y })

  onMapClick = (info, e) => {
    if (info.objects) {
      if (this.state.viewport.zoom >= 20 && this.state.hideDrawer) {
        this.setState({ hideDrawer: false })
      } else {
        // Zoom to center of cluster
        const features = featureCollection(info.objects.map(o => point([o.coordinates[0], o.coordinates[1]])))
        const c = center(features)

        this.setState({ viewport: {
          ...this.state.viewport,
          longitude: c.geometry.coordinates[0],
          latitude: c.geometry.coordinates[1],
          // Don't zoom out if already further in than expansionZoom
          zoom: Math.max(this.state.viewport.zoom, info.expansionZoom),
          transitionDuration: 500,
          transitionInterpolator: new FlyToInterpolator()
        } })
      }
    } else {
      this.setState({ selectedObject: info.object })
      if (info.object.type === 'member') {
        this.props.gotoMember(info.object.id)
      } else if (info.object.type === 'community') {
        this.props.showCommunityDetails(info.object.id)
      } else {
        this.props.showDetails(info.object.id)
      }
    }
  }

  toggleFeatureType = (type, checked) => {
    const featureTypes = this.props.filters.featureTypes
    featureTypes[type] = checked
    this.props.storeClientFilterParams({ featureTypes })
  }

  _renderTooltip = () => {
    const { hoveredObject, pointerX, pointerY } = this.state || {}
    if (hoveredObject) {
      let message
      let type
      if (Array.isArray(hoveredObject) && hoveredObject.length > 0) {
        // cluster
        const types = groupBy(hoveredObject, 'type')
        message = Object.keys(types).map(type => <p key={type}>{types[type].length} {type}{types[type].length === 1 ? '' : 's'}</p>)
        type = 'cluster'
      } else {
        message = hoveredObject.message
        type = hoveredObject.type
      }

      return (
        <div styleName='postTip' className={styles[type]} style={{ left: pointerX + 15, top: pointerY }}>
          { message }
        </div>
      )
    }
    return ''
  }

  toggleDrawer = (e) => {
    this.setState({ hideDrawer: !this.state.hideDrawer })
    this.props.toggleDrawer(!this.state.hideDrawer)
  }

  toggleFeatureFilters = (e) => {
    this.setState({ showFeatureFilters: !this.state.showFeatureFilters })
  }

  toggleSavedSearches = (e) => {
    this.setState({ showSavedSearches: !this.state.showSavedSearches })
  }

  saveSearch = (name) => {
    const { boundingBox } = this.state
    const { currentUser, filters, location, posts } = this.props
    const { featureTypes, search: searchText, topics } = filters
    const { pathname } = location;

    let context
    if (isAllCommunitiesPath(pathname)) context = 'all'
    if (isPublicPath(pathname)) context = 'public'
    if (isCommunityPath(pathname)) context = 'community'
    if (isNetworkPath(pathname)) context = 'network'

    let communitySlug
    let networkSlug
    
    if (context === 'community') communitySlug = getCommunitySlug(pathname)
    if (context === 'network') networkSlug = getNetworkSlug(pathname)

    const userId = currentUser.id

    const postTypes = Object.keys(featureTypes).reduce((selected, type) => {
      if (featureTypes[type]) selected.push(type)
      return selected;
    }, [])

    const lastPostId = get(posts, '[0].id')

    const topicIds = topics.map(t => t.id)

    const attributes = { boundingBox, communitySlug, context, lastPostId, name, networkSlug, postTypes, searchText, topicIds, userId }

    this.props.saveSearch(attributes)
  }

  render () {
    const {
      currentUser,
      fetchParams,
      deleteSearch,
      features,
      filters,
      pending,
      routeParams,
      searches,
      topics
    } = this.props

    const {
      clusterLayer,
      communityIconLayer,
      features,
      hideDrawer,
      showFeatureFilters,
      showSavedSearches,
      viewport
    } = this.state

    return <div styleName={cx('container', { 'noUser': !currentUser })}>
      <div styleName='mapContainer'>
        <Map
          layers={[communityIconLayer, clusterLayer]}
          afterViewportUpdate={this.afterViewportUpdate}
          onViewportUpdate={this.mapViewPortUpdate}
          children={this._renderTooltip()}
          viewport={viewport}
        />
        { pending && <Loading className={styles.loading} /> }
      </div>
      <button styleName={cx('toggleDrawerButton', { 'drawerOpen': !hideDrawer })} onClick={this.toggleDrawer}>
        <Icon name='Hamburger' className={styles.openDrawer} />
        <Icon name='Ex' className={styles.closeDrawer} />
      </button>
      { !hideDrawer ? (
        <MapDrawer
          currentUser={currentUser}
          features={features}
          fetchParams={fetchParams}
          filters={filters}
          onUpdateFilters={this.props.storeClientFilterParams}
          pending={pending}
          routeParams={routeParams}
          topics={topics}
        />)
        : '' }
      <div styleName={cx('searchAutocomplete')}>
        <LocationInput saveLocationToDB={false} onChange={(value) => this.handleLocationInputSelection(value)} />
      </div>
      <button styleName={cx('toggleFeatureFiltersButton', { 'featureFiltersOpen': showFeatureFilters })} onClick={this.toggleFeatureFilters}>
        Post Types: <strong>{Object.keys(filters.featureTypes).filter(t => filters.featureTypes[t]).length}/5</strong>
      </button>
      <img styleName={cx('savedSearchesButton')} onClick={this.toggleSavedSearches} src={heart} />
      { showSavedSearches ? (<SavedSearches deleteSearch={deleteSearch} filters={filters} saveSearch={this.saveSearch} searches={searches} toggle={this.toggleSavedSearches} />) : '' }
      <div styleName={cx('featureTypeFilters', { 'featureFiltersOpen': showFeatureFilters })}>
        <h3>What do you want to see on the map?</h3>
        {['member', 'request', 'offer', 'resource', 'event'].map(featureType => {
          let color = FEATURE_TYPES[featureType].primaryColor

          return <div
            key={featureType}
            ref={this.refs[featureType]}
            styleName='featureTypeSwitch'
          >
            <SwitchStyled
              backgroundColor={`rgba(${color[0]}, ${color[1]}, ${color[2]}, ${color[3] / 255})`}
              name={featureType}
              checked={filters.featureTypes[featureType]}
              onChange={(checked, name) => this.toggleFeatureType(name, !checked)}
            />
            <span>{featureType.charAt(0).toUpperCase() + featureType.slice(1)}s</span>
          </div>
        })}
        <div styleName={cx('pointer')} />
      </div>
    </div>
  }
}
