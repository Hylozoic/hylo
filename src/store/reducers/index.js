import { combineReducers } from 'redux'
import { connectRouter } from 'connected-react-router'
import { history } from 'router'

import orm from './ormReducer'
import returnToPath from 'store/reducers/returnToPath'
import graphqlResponseError from 'store/reducers/graphqlResponseError'
import pending from './pending'
import locationHistory from './locationHistory'
import resetStore from './resetStore'
import mixpanel from './mixpanel'
import intercom from './intercom'
import queryResults from './queryResults'
import { handleSetState, composeReducers } from './util'

// Local store
// generator-marker-local-store-import
import AllTopics from 'routes/AllTopics/AllTopics.store'
import AttachmentManager from 'components/AttachmentManager/AttachmentManager.store'
import CreateGroup from 'components/CreateGroup/CreateGroup.store'
import CreateTopic from 'components/CreateTopic/CreateTopic.store'
import Events from 'routes/Events/Events.store'
import FeedList from 'components/FeedList/FeedList.store'
import FullPageModal from 'routes/FullPageModal/FullPageModal.store'
import HyloEditor from 'components/HyloEditor/HyloEditor.store'
import MapExplorer from 'routes/MapExplorer/MapExplorer.store'
import Messages from 'routes/Messages/Messages.store'
import Members from 'routes/Members/Members.store'
import MembershipRequests from 'routes/GroupSettings/MembershipRequestsTab/MembershipRequestsTab.store'
import MemberSelector from 'components/MemberSelector/MemberSelector.store'
import ModeratorsSettings from 'routes/GroupSettings/ModeratorsSettingsTab/ModeratorsSettingsTab.store'
import PeopleTyping from 'components/PeopleTyping/PeopleTyping.store'
import PostEditor from 'components/PostEditor/PostEditor.store'
import AuthLayoutRouter from 'routes/AuthLayoutRouter/AuthLayoutRouter.store'
import RelatedGroups from 'routes/GroupSettings/RelatedGroupsTab/RelatedGroupsTab.store'
import SavedSearches from 'routes/UserSettings/UserSettings.store'
import Search from 'routes/Search/Search.store'
import SkillsSection from 'components/SkillsSection/SkillsSection.store'
import SkillsToLearnSection from 'components/SkillsToLearnSection/SkillsToLearnSection.store'
import TopicsSettings from 'routes/GroupSettings/TopicsSettingsTab/TopicsSettingsTab.store'
import UserGroupsTab from 'routes/UserSettings/UserGroupsTab/UserGroupsTab.store'

export const combinedReducers = combineReducers({
  // Global store
  orm,
  router: connectRouter(history),
  returnToPath,
  graphqlResponseError,
  pending,
  queryResults,
  locationHistory,
  mixpanel,
  intercom,

  // Local store (Component)
  // generator-marker-local-store-reducer
  AllTopics,
  AttachmentManager,
  CreateGroup,
  CreateTopic,
  Events,
  FeedList,
  FullPageModal,
  HyloEditor,
  MembershipRequests,
  MapExplorer,
  Members,
  MemberSelector,
  Messages,
  ModeratorsSettings,
  PeopleTyping,
  AuthLayoutRouter,
  PostEditor,
  RelatedGroups,
  SavedSearches,
  Search,
  SkillsSection,
  SkillsToLearnSection,
  TopicsSettings,
  UserGroupsTab
})

export default composeReducers(
  combinedReducers,
  resetStore,
  handleSetState
)
