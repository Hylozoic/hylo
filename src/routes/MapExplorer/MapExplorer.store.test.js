import { fetchPosts } from './MapExplorer.store'

it('works for a group', () => {
  expect(fetchPosts({
    context: 'groups',
    id: 'foo',
    offset: 20,
    search: 'gardening',
    filter: 'offer'
  })).toMatchSnapshot()
})

it('works for all groups', () => {
  expect(fetchPosts({
    context: 'all',
    offset: 20,
    search: 'graphic design',
    filter: 'request'
  })).toMatchSnapshot()
})
