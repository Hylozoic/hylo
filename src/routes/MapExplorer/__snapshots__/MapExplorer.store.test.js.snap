// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works for a community 1`] = `
Object {
  "graphql": Object {
    "query": "query (
  $slug: String,
  $sortBy: String,
  $filter: String,
  $search: String,
  $topic: ID,
  $first: Int,
  $offset: Int,
  $boundingBox: [PointInput]
) {
  community(slug: $slug, updateLastViewed: true) {
    id
    slug
    name
    avatarUrl
    bannerUrl
    postCount
    
posts(
  first: $first,
  offset: $offset,
  sortBy: $sortBy,
  search: $search,
  filter: $filter,
  topic: $topic,
  boundingBox: $boundingBox,
  order: \\"desc\\"
) {
  hasMore
  items {
    
  id
  announcement
  title
  details
  type
  creator {
    id
    name
    avatarUrl
  }
  createdAt
  updatedAt
  isPublic
  fulfilledAt
  startTime
  endTime
  myEventResponse
  commenters(first: 3) {
    id
    name
    avatarUrl
  }
  commentersTotal
  
  linkPreview {
    id
    title
    url
    imageUrl
  }
  location
  locationObject {
    id
    addressNumber
    addressStreet
    bbox {
      lat
      lng
    }
    center {
      lat
      lng
    }
    city
    country
    fullText
    locality
    neighborhood
    region
  }
  votesTotal
  myVote
  acceptContributions
  totalContributions
  communities {
    id
    name
    slug
  }
  attachments {
    type
    url
    position
    id
  }
  postMemberships {
    id
    pinned
    community {
      id
    }
  }
  topics {
    id
    name
    postsTotal
    followersTotal
  }
  members {
    total
    hasMore
    items {
      id
      name
      avatarUrl
      bio
      tagline
      location
      skills (first: 100) {
        items {
          id
          name
        }
      }
    }
  }
  eventInvitations {
    total
    hasMore
    items {
      id
      response
      person {
        id
        name
        avatarUrl
        bio
        tagline
        location
        skills (first: 100) {
          items {
            id
            name
          }
        }
      }
    }
  }
  }
}
  }
}",
    "variables": Object {
      "boundingBox": undefined,
      "filter": "offer",
      "networkSlug": undefined,
      "search": "gardening",
      "slug": undefined,
      "sortBy": undefined,
      "topic": undefined,
    },
  },
  "meta": Object {
    "extractModel": "Community",
    "extractQueryResults": Object {
      "getItems": [Function],
    },
  },
  "type": "FETCH_POSTS_MAP",
}
`;

exports[`works for all communities 1`] = `
Object {
  "graphql": Object {
    "query": "query (
  $sortBy: String,
  $search: String,
  $filter: String,
  $topic: ID,
  $first: Int,
  $offset: Int,
  $boundingBox: [PointInput]
) {
  
posts(
  first: $first,
  offset: $offset,
  sortBy: $sortBy,
  search: $search,
  filter: $filter,
  topic: $topic,
  boundingBox: $boundingBox,
  order: \\"desc\\"
) {
  hasMore
  items {
    
  id
  announcement
  title
  details
  type
  creator {
    id
    name
    avatarUrl
  }
  createdAt
  updatedAt
  isPublic
  fulfilledAt
  startTime
  endTime
  myEventResponse
  commenters(first: 3) {
    id
    name
    avatarUrl
  }
  commentersTotal
  
  linkPreview {
    id
    title
    url
    imageUrl
  }
  location
  locationObject {
    id
    addressNumber
    addressStreet
    bbox {
      lat
      lng
    }
    center {
      lat
      lng
    }
    city
    country
    fullText
    locality
    neighborhood
    region
  }
  votesTotal
  myVote
  acceptContributions
  totalContributions
  communities {
    id
    name
    slug
  }
  attachments {
    type
    url
    position
    id
  }
  postMemberships {
    id
    pinned
    community {
      id
    }
  }
  topics {
    id
    name
    postsTotal
    followersTotal
  }
  members {
    total
    hasMore
    items {
      id
      name
      avatarUrl
      bio
      tagline
      location
      skills (first: 100) {
        items {
          id
          name
        }
      }
    }
  }
  eventInvitations {
    total
    hasMore
    items {
      id
      response
      person {
        id
        name
        avatarUrl
        bio
        tagline
        location
        skills (first: 100) {
          items {
            id
            name
          }
        }
      }
    }
  }
  }
}
}",
    "variables": Object {
      "boundingBox": undefined,
      "filter": "request",
      "networkSlug": undefined,
      "search": "graphic design",
      "slug": undefined,
      "sortBy": undefined,
      "topic": undefined,
    },
  },
  "meta": Object {
    "extractModel": "Post",
    "extractQueryResults": Object {
      "getItems": [Function],
    },
  },
  "type": "FETCH_POSTS_MAP",
}
`;
